Ledger API (#16)
hyperledger/fabric-rfcs: text/0000-ledger-api.md

---
layout: default
title: Ledger_API
parent: RFCs
---

- Feature Name: Ledger_API
- Start Date: 2019-12-03
- RFC PR: (leave this empty)
- Fabric Issue: (leave this empty)

# Summary
[summary]: #summary

> This feature is a continuation of the updates made in recent releases to the Fabric Programming Model. This RFC's aim is to clarify the abstractions used to work with ledger data in a contract. It's scope is restricted to what is today referred to as the 'ChaincodeStub API', and specifically at the APIs around states.

Reviewer: ```suggestion
This feature is a continuation of the updates made in recent releases to the Fabric Programming Model. This RFC's aim is to clarify the abstractions used to work with ledger data in a contract. It's scope is restricted to what is today referred to as the 'ChaincodeStub API'-- specifically  the APIs around states.
```

This is a cross-language RFC for Golang, JavaScript, TypeScript and Java.

# Motivation
[motivation]: #motivation

> The updated Programming Model, has introduced higher level abstractions to both the contracts and the client SDKs. Many users had already written high-level utility or support code within their own applications. Abstractions help us think about and solve problems, improving usability, speed of development and performance.

Reviewer: ```suggestion
The updated Programming Model has introduced higher level abstractions to both the contract and the client SDKs. Many users have already written high-level utility or support code within their own applications. Abstractions help us think about and solve problems, improving usability, speed of development and performance.
```

> Examples, such as the Commercial Paper tutorial, also take this approach. It simplifies the concepts and makes working with business level data simpler. (c.f. https://jira.hyperledger.org/browse/FABCN-42)

Reviewer: ```suggestion
Examples, such as the Commercial Paper tutorial, also take this approach. It simplifies the concepts and makes working with business-level data simpler. (c.f. https://jira.hyperledger.org/browse/FABCN-42)
```

> This RFC (which is anticipated is the first of a series) will formalize the existing 'stub-api' into a conceptual architecture. A key aspect in all the programming model work is that the conceptual model is consistent between languages. The polyglot developer will therefore have minimal friction moving between languages.  Note that this a *conceptual* model; each implementation will use idiomatic language code, and where appropriate, specifics of the language to give the best developer experience.

Reviewer: ```suggestion
This RFC (which is anticipated as the first in a series) will formalize the existing 'stub-api' into a conceptual architecture. A key aspect in all the programming model work is that the conceptual model is consistent between languages. The polyglot developer will therefore have minimal friction moving between languages.  Note that this a *conceptual* model; each implementation will use idiomatic language code, and where appropriate, specifics of the language to provide the best developer experience.
```

> Across all the languages today, the stub API is broadly consistent between languages. However there are variations that are confusing when looking across languages. Within each language as well there are APIs that have minor inconsistencies. Whilst not major, these differences are sufficient to cause a developer uncertainty in how an API should be used. Introducing a ledger formalizes and integrates existing concept..

Reviewer: ```suggestion
The stub API is broadly consistent across languages. There are, however, variations and minor inconsistencies that cause confusion when looking across languages. Whilst not major, these differences are sufficient to cause a developer uncertainty in how an API should be used. Introducing a ledger formalizes and integrates existing concepts.
```

Future RFCs will focus on the persistence of the business data; many of the same helper classes used today are based around the idea of serializing data, and move the knowledge of key creation into the business objects themselves. (c.f. https://jira.hyperledger.org/browse/FABCN-92)

# Guide-level explanation
[guide-level-explanation]: #guide-level-explanation

A complete solution using Fabric can be represented as follows:
![](0000-ledger-api-ConceptualArchitecture.png)

> A typical sequence is for an application to connect to a peer (the 'gateway' into the network channel), and obtain a reference to a smart contract. The Network-shared transaction logic in the Smart Contract is then be invoked by name, and with optional data arguments and return values. The data arguments can include transient data if needed. The logic in the Smart Contract will typical reference and update the shared transaction data, subject of course to exactly how the code is written.

Reviewer: ```suggestion
A typical sequence is for an application to connect to a peer (the 'gateway' into the network channel), and obtain a reference to a smart contract. The Network-shared transaction logic in the Smart Contract is then be invoked by name with optional data arguments and return values. The data arguments can include transient data if needed. The logic in the Smart Contract will typical reference and update the shared transaction data, subject of course to exactly how the code is written.
```

The 'Ledger' is, therefore, a primary abstraction in a blockchain system, including in Hyperledger Fabric; this abstraction allows the concepts of states, world state, collections, transactions, blocks etc. to be tied together and formalised. 

Looking at the objects that make up this conceptual architecture, a full set of objects would have the following relationships. 

![](0000-ledger-api-relationships.png)

> Note that this RFCs is concerned with the ledger-api and the ledger,collection and state; the 'white' boxes will be subsequent RFCs.

Reviewer: ```suggestion
Note that this RFC is concerned with the ledger-api and the ledger,collection and state; the 'white' boxes will be subsequent RFCs.
```

This is a cross-language RFC for Golang, JavaScript, TypeScript and Java.

# Motivation
[motivation]: #motivation

The updated Programming Model has introduced higher level abstractions to both the contract and the client SDKs. Many users have already written high-level utility or support code within their own applications. Abstractions help us think about and solve problems, improving usability, speed of development and performance.

Examples, such as the Commercial Paper tutorial, also take this approach. It simplifies the concepts and makes working with business-level data simpler. (c.f. https://jira.hyperledger.org/browse/FABCN-42)

This RFC (which is anticipated as the first in a series) will formalize the existing 'stub-api' into a conceptual architecture. A key aspect in all the programming model work is that the conceptual model is consistent between languages. The polyglot developer will therefore have minimal friction moving between languages.  Note that this a *conceptual* model; each implementation will use idiomatic language code, and where appropriate, specifics of the language to give the best developer experience.

> The stub API is broadly consistent between languages. There are, however, variations and minor inconsistencies that cause confusion when looking across languages. Whilst not major, these differences are sufficient to cause a developer uncertainty in how an API should be used. Introducing a ledger formalizes and integrates existing concepts. 

Reviewer 2: Curious to know about these inconsistencies. An example would be better.

Author: I've also added some links to JIRAs later in the RFC.

A lot of these are minor in the spelling of the APIs varies between languages. 
The return type of the iterators is some APIs in node.js wrapped in other objects. 
APIs in Java have been missed out for example.

> To allow for functional differences between different but very similar operations the stubAPI has chosen an approach to have a 'broad-and-flat' approach to APIs. For example there are query APIs per type of collection, `getPrivateDataByRange` and `getStateByRange`, by apply modifiers to the function name the function is extended. The length of the resulting list of APIs is larger and should additional function become available will increase. A long list of APIs increases the difficulty of finding the API needed, especially without autocomplete-style tool support. By formalizing the APIs in the proposed structure makes an overall more sustainable model to support future changes.  

Reviewer 2: why _getPrivateDataByRange_ and _getStateByRange_ are referred to as the API accessing two different types of collections. By collection/private data, we store the private data locally on the peer (as per the membership) and store only the hash of the private data (as proof of the existence of data) on the blockchain. The world/public state refers to data present in the blocks -- which are public data and hash of private data.

Author: Correct yes at a Technical Fabric level; From a developer of the Smart Contract, the way they access the data follows the same semantics. Hence allowing the same style of API to access them - but what they are doing 'underneath' is as you state.

The 'Ledger' is, therefore, a primary abstraction in a blockchain system, including in Hyperledger Fabric; this abstraction allows the concepts of states, world state, collections, transactions, blocks etc. to be tied together and formalised. 

Looking at the objects that make up this conceptual architecture, a full set of objects would have the following relationships. 

![](0000-ledger-api-relationships.png)

Note that this RFC is concerned with the ledger-api and the ledger,collection and state; the 'white' boxes will be subsequent RFCs.

## Ledger

A Ledger's current value is held in a set of collections. A single ledger is available to any given deployed contract, though other ledgers may be available within the Network Channel and can be accessed via other contracts

> Of the collections that are reference by the ledger, there a single collection in the ledger that is public (the 'WorldState') and held by every organization in the Network Channel. A ledger may have 1 or more private collections that are available only in some organizations. Starting at Fabric v2.0 there will be an implicit private collection per organization.

Reviewer 2: Not sure whether we can call both the world state and private state as a collection. As far as I know, the term collection has been used only for private data. By collection, I always thought that the actual data gets into a collection and the hash of the data gets into the block. Whatever data present in the block is called the world state is what I understood. Please correct me if I have misunderstood. 

I remember that one of the usages of the collection is to not leak the data to ordering service by creating a collection with all members. I thought this to be the public collection.

Author: Your understanding of Fabric is correct - and this API is not changing anything about the concept. 
Today for example if you look at putState() and putPrivateState() there are different APIs, but in terms of the gRPC flows to the peer - it's the same low-level flow.. the gRPC level-putstate, takes a string, which is the empty string for the worldstate, and a name for private data.

## Collection

A collection is a set of states, with each state holding a business object or data. Each state being addressed by a key. Private collections are identified by name, and held within a set of policy-defined organizations.

Accessing any given collection is via the Ledger instance. 

```java
	Ledger = Ledger.getLedger(ctx)

	// predefined constant for the world state
	Collection worldState = Ledger.getCollection(Collection.WORLD);

	// Private Data collections are reference by name
	Collection privateData = Ledger.getCollection("PrivateSalesData");

	// Each organization will have an implicit private data collection created for them
	// this has a prefixed name
	Collection org1PrivateData = Ledger.getCollection(Collection.organizationCollection("orgsmspID"));

	// with a convience version of
	Collection org1PrivateData = Ledger.getOrganizationCollection("orgsmspID");

```

There are underlying technical restrictions on some operations. For example you cannot combine private data range queries and updates in a transaction.
Nor do private data collections support pagination. Should these not be supported, a 'not-supported' error/exception will be thrown.

> As there are two essential types of collection - public and private - this is available via an enumerated value. Each languages approach to enumerations is syntactical different but conceptually similar

Reviewer 2: Some of the comment that I mentioned above is applicable here.

Author: It is true that the word collection is being overloaded?  Is there an alternative you can think of?

Reviewer 2: We are trying to put the following three variants under a single bucket called a collection. I am completely fine with that provided that it does not create any confusion for the developers. 

1. **Private collections**: only a sub-set of channel members are part of this collection. It provides data privacy within a channel. 
2. **Public collections**: everyone in the channel is a member of this collection. This allows us to have different endorsement policy for public data (coarse-granular compared to a key-based endorsement but fine-granular compared to chaincode level endorsement policy). Further, it helps in avoiding the data leak at the ordering service, for example, if the deployment uses a thrid-party ordering service (provided by some cloud vendors). Note that this concept is not explicitly discussed in Fabric but we can create such a collection even now as the only difference compared to a private collection is the member list. We did discuss the concept of the public collection while designing the private data but didn't write it on any docs. In one of the comments above, Dave mentioned about bringing the public collection concept to core fabric. 
3. **World state**: States that are visible to everyone in the channel including the ordering service. This is the state that actually presents in the blocks. We can also term this as blockchain states or ledger data. This includes the hash of the private data which can be accessed by everyone in the channel through `GetPrivateDataHash()`. Further, a few patterns have been introduced by Dave using this new API which makes the developer aware of hash of some private data belonging to a sub-set of organizations is present in the blockchain.  

IMO, public and private collection can be called a collection without the actual differentiation as the underlying concept is the same (these data do not get into blocks) but there is a slight difference in the goal. Given that, we might call the world state as public collection and the other as a private collection. Again, I am not sure whether it would confuse the developers.

```
	COLLECTION_TYPE = {PUBLIC , PRIVATE}
```

## State

A state holds the value of a business object or data, addressed by a key. The format of this object or data is currently defined by the overall solution, and encoding the format of 'the value' of the state is the responsibility of the contract developer.

The state can also anchor the following functions
 - Access to individual states can be controlled by state-level endorsement policies, APIs are provided to assist in the construction and parsing of these policies. 
 - Access to the Private Data Hash - essential for private data scenarios.
 - Access to history details of the state, what transactions affect this state in the past
 - Creation of composite keys and general key handling helpers

## Example Interfaces

To see how the concepts above can be rendered in code we can walk-through an updated version of the Java Commercial Paper example.

### Putting data

The 'issue' method will create a new business object based on supplied data, and store this within the Ledger.

> ```java

Reviewer 2: **Proposed Way**
 ```
Collection collection = Ledger.getLedger(ctx).getCollection(Collection.WORLD);
String key = State.makeComposite(CommercialPaper.class.getName(), new String[] { paperNumber });
collection.createState(key, CommercialPaper.serialize(paper));
```
**Old way**
```
String key = CreateCompositeKey(CommercialPaper.class.getName(), new String[] { paperNumber });
PutState(key, CommercialPaper.serialize(paper));
```
I would like to understand how the proposed way is better than the old way of writing contracts. After reading further, I understand that if the user wants to store the paper on a private collection later, just changing one line is adequate. Does it happen often? Isn't the old way look simpler (as we have to type a few characters) and readable? It may be subjective. 

Is the motivation of this proposal to reduce the number of chaincode interfaces to simplify the task of chaincode developers? I am asking this because, with this proposal, the 9 APIs dedicated to the private data would not be needed. Given that, I would like to state the reasoning for defining separate APIs for private data. The following are some of the points copied from an email chain dated July 2017 that decided to use separate APIs for private data. 

- Is private data really part of ledger state? The way we currently manage it (or propose to manage) is outside of the ledger. Its hash is on the ledger, so its hash is part of the state but the actual data is not. So I think we should be very clear to the developers that this is outside of the blockchain. IMO, the best would be to express this intention with completely different APIs than get/putState.  Perhaps stub.get/putPriviateData(...)
- exposing the same APIs for public and private data may be a bit confusing for the users. 
- Also, the use of collections in the public data is not clear as such. But in future, if we find any use of collections in the public data, we would be back to the same problem - how to differentiate public data and private data. This would demand again a change in the APIs.
